docs/proposals/project-repos-and-clusters.md

```diff
@@ -0,0 +1,154 @@
+---
+title: Neat-enhancement-idea
+authors:
+- "@alexmt"
+  sponsors:
+- TBD
+  reviewers:
+- "@jessesuen"
+- TBD
+  approvers:
+- "@jessesuen"
+- TBD
+
+creation-date: 2020-04-19
+last-updated: 2020-04-19
+---
+
+# Neat Enhancement Idea
+
+Support project scoped Repositories and Clusters to enable self-service end-users onboarding.
+
+## Summary
+
+The Argo CD has two type of users:
+
+* Administrators who configure the Argo CD and manage the Argo CD projects.
+* Developers who use Argo CD to manage resources in the Kubernetes clusters.
+
+These two roles enable sharing on the Argo CD instance in a multi-tenant environment. Typically the developer
+requests a new project from an administrator. The administrator creates the project, defines which repositories
+can and clusters can be used within the project which concludes the onboarding.
+
+The problem is that list of repositories and clusters often not known during the onboarding process. Developers get
+it later and have to again contact an administrator, somehow share repo/cluster credentials. This back and forth
+process takes time and creates friqtion.
+
+We want to streamline the process of adding repositories and clusters to the project and make it self-service.
+
+## Motivation
+
+As long as the developer has the required credentials he/she should be able to add repository/cluster to the project
+without involving the administrator. To archive it, we are proposing to introduce project scoped repositories and clusters.
+
+### Goals
+
+The goals of project scoped repositories and clusters are:
+
+#### Allow Self-Registering Repositories/Clusters in a Project
+
+Developer should be able to add a repository/cluster into the project without asking help from Argo CD administrator.
+
+### Non-Goals
+
+#### Simplify Management of Shared Repositories/Clusters in a Project
+
+The repositories and clusters that can be used across multiple projects still have to be managed by Argo CD administrator.
```

Comment:
Adding clusters across multiple projects is one use-case I am interested in. Is this a non-goal of this proposal because of the proposed implementation or another reason?

With this proposal adding repositories/clusters to multiple projects fails because the primary key already exists,
the primary key of project scoped entities could extended to include the project name to allow `project1-cluster-a`
and `project2-cluster-a`, while still allowing backwards compatibility with the un-scoped `cluster-a` resources.

Reply:
It is pretty expensive to "monitor" a cluster: Argo CD has to keep one open connection per resource in a cluster and store some information in about resources in memory.

So even if two teams provide cluster credentials we would have to pick one. It is not clear how to decide automatically which credentials should be used.

Reply:
To summarize: the proposal describes the first step that allows to self-onboard non-shared repos/clusters. 
I would be glad to support shared ones but it is not clear how to do it